compiler/rustc_mir_build/src/build/mod.rs

```diff
@@ -740,6 +750,7 @@ fn construct_error<'a, 'tcx>(hir: Cx<'a, 'tcx>, body_id: hir::BodyId) -> Body<'t
 impl<'a, 'tcx> Builder<'a, 'tcx> {
     fn new(
         hir: Cx<'a, 'tcx>,
+        def_id: DefId,
```

Comment:
```suggestion
        def: ty::WithOptConstParam<DefId>,
```
I think it should be easier to already pass the correct `WithOptConstParam` into this builder.

Reply:
That's the part I'm not understanding. What are the invariants of `WithOptConstParam`? When is it okay to be `unknown`? Do I have to call `try_upgrade` manually?

Reply:
The invariant of `WithOptConstParam` is that if `self.const_param_did.is_some()`, then `self.const_param_did == tcx.opt_const_param_of(self.did)`. To prevent cycle errors `self.const_param_did` should also be `Some` if it is used during `typeck`, `did` is a const argument and we want to know `tcx.type_of(did)`, which we then replace with `tcx.type_of(def.def_id_for_type_of())`.

You should never have to call `try_upgrade` except at the start of a query taking `ty::WithOptConstParam` so that we only evaluate it once.

It is okay to use `unknown` if we are either done with typeck or `tcx.opt_const_param_of(self.did)` would/did return `None`.

Looks like it might be worth it to add a section to the rustc-dev-guide about `WithOptConstParam` :thinking:

Reply:
> Looks like it might be worth it to add a section to the rustc-dev-guide about WithOptConstParam thinking

As I've alluded to, I think time would be better spent better encapsulating this system. I don't particularly want to think about it when writing MIR transformations. I realize I'm being a bit curmudgeonly here and that this is in service of a feature which I'm super excited about (thanks BTW!), but technical debt is a real concern, especially in mir-opt land.

Is there anything we could do to reduce the surface area here? We can discuss this in a separate issue I guess.

Reply:
Looks like I didn't send the message I wanted to send here :sweat_smile: 

While working on this I experimented with some different approaches and the less invasive ones all ended up being unsound in the face of incremental compilation. So I would be glad - and a bit annoyed :laughing: - if you are able to come up with a less invasive solution here, but do think that this will be quite hard to do

Reply:
I have some ideas, but if you've already explored alternatives, it's unlikely that I'll be able to do better. I've [not given up quite yet](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/Alternatives.20to.20.60ty.3A.3AWithOptConstParam.60) though :smile:.